substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -672,11 +674,28 @@ pub mod pallet {
 
             log::info!(
                 "{:?} contracts to bill at block {:?}",
-                contracts,
+                contract_ids,
                 block_number
             );
 
-            for contract_id in contracts {
+            for contract_id in contract_ids {
+                if let Some(c) = Contracts::<T>::get(contract_id) {
+                    if let types::ContractData::NodeContract(node_contract) = c.contract_type {
+                        let bill_ip = node_contract.public_ips > 0;
+                        let bill_cu_su = NodeContractResources::<T>::contains_key(contract_id)
+                            && !NodeContractResources::<T>::get(contract_id).used.is_empty();
+                        let bill_nu =
+                            ContractBillingInformationByID::<T>::contains_key(contract_id)
+                                && ContractBillingInformationByID::<T>::get(contract_id)
+                                    .amount_unbilled
+                                    > 0;
+
+                        // Don't bill if no IP/CU/SU/NU to bill
+                        if !bill_ip && !bill_cu_su && !bill_nu {
+                            continue;
+                        }
+                    }
+                }
                 let _res = Self::bill_contract_using_signed_transaction(contract_id);
```

Comment:
move this one line up? If there is no contract (deleted somehow but still in billing loop) we don't want to push an extrinsic?

Reply:
Hmm, lets check https://github.com/threefoldtech/tfchain/issues/703#issuecomment-1560631473 for this

Reply:
> move this one line up? If there is no contract (deleted somehow but still in billing loop) we don't want to push an extrinsic?

cannot move this line up because when contract is deleted its id in billing loop is cleaned up at next cycle by pushing the extrinsic

Reply:
Yes I just noticed that as well this morning, but it's very weird and hidden behaviour. It's not clear that it's subject to the offchain worker to be cleaned up or not.